### ISSUE TITLE
self.scaler.get_scale() forces cpu and gpu to synchronize

### ISSUE THREAD

**Issue by StarCycle** [huggingface/accelerate#1714]
### System Info

https://github.com/huggingface/accelerate/blob/c33adecc9f92808cbaeb0d34928e52e34180f964/src/accelerate/optimizer.py#L125C32-L125C55

In this line, "scale_before = self.scaler.get_scale()" will force a synchronization between CPU and GPU, and may make the training process slower. You can see this behaviour in pytorch profiler

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [ ] My own task or dataset (give details below)

### Reproduction

1. Run the following code in single gpu without mixed precision, and read the profile trace with tensorboard or chrome, you can see the cpu and gpu does not need to synchronize (Dataloader will do cpu data prefetching and DataPrefetcher will do gpu data prefetching).
2. Run the following code in single gpu with mixed precision. In the trace, you can see the cpu and gpu must synchronize before updating the network weights, because Accelerate calls "scale_before = self.scaler.get_scale()" and "scale_after = self.scaler.get_scale()" in optimizer.py.
3. Disable the 2 lines and run the following again, you can see the cpu and gpu does NOT need to synchronize and the GPU utilization becomes higher.

```python
import torch
import torchvision
import torchvision.transforms as T
from torch.profiler import profile, record_function, ProfilerActivity, tensorboard_trace_handler, schedule

def train(inputs, labels, scaler):
    with torch.autocast(device_type='cuda', dtype=torch.float16):
        outputs = model(inputs)
        loss = criterion(outputs, labels)
    optimizer.zero_grad()
    # loss.backward()
    # optimizer.step()
    scaler.scale(loss).backward()
    scaler.step(optimizer)
    scaler.update()

class DataPrefetcher():
    def __init__(self, loader, device):
        self.device = device
        self.loader = iter(loader)
        self.stream = torch.cuda.Stream()
        self.preload()

    def preload(self):
        try:
            self.next_input, self.next_label = next(self.loader)
        except StopIteration:
            self.next_input = None
            self.next_label = None
            return
        with torch.cuda.stream(self.stream):
            self.next_input = self.next_input.to(device, non_blocking=True)
            self.next_label = self.next_label.to(device, non_blocking=True)

    def next(self):
        torch.cuda.current_stream().wait_stream(self.stream)
        input = self.next_input
        label = self.next_label
        if input is not None:
            input.record_stream(torch.cuda.current_stream())
        if label is not None:
            label.record_stream(torch.cuda.current_stream())
        self.preload()
        return input, label

torch.backends.cuda.matmul.allow_tf32 = True
torch.backends.cudnn.allow_tf32 = True
torch.backends.cudnn.benchmark = True
device = torch.device('cuda:1')
transform = T.Compose(
    [T.Resize(224),
     T.ToTensor(),
     T.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
train_set = torchvision.datasets.CIFAR10(root='./data', train=True, download=True, transform=transform)
loader = torch.utils.data.DataLoader(train_set, batch_size=208, shuffle=True, num_workers=32, pin_memory=True)
prefetcher = DataPrefetcher(loader, device)
model = torchvision.models.swin_v2_t(weights='IMAGENET1K_V1').to(device)
criterion = torch.nn.CrossEntropyLoss().to(device)
optimizer = torch.optim.AdamW(model.parameters(), lr=1e-5, weight_decay=1e-4, fused=True)
scaler = torch.cuda.amp.GradScaler()
model.train()
prof = profile(
        activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA],
        schedule=schedule(wait=10, warmup=3, active=2),
        on_trace_ready=tensorboard_trace_handler('./log'),
        record_shapes=True,
        profile_memory=True,
        with_stack=True,
        with_flops=True,
        with_modules = True,
)
prof.start()
input, label = prefetcher.next()
iteration = 0
while input is not None:
    if iteration >= 10000:
        # prof.stop()
        break
    train(input, label, scaler)
    # prof.step()  
    iteration += 1
    input, label = prefetcher.next()

```
### Expected behavior

As mentioned above.

**Comment by muellerzr** [comment]
Thanks for the detailed report here, yes indeed in their docs they mention this will force a CPU and GPU sync-up. This checks for overflow in the optimizer during mixed precision, so we can't just get rid of it entirely. We're working on a framework inside accelerate for checking shapes and such, and there will be a hyper-optimized version for speed (going without any guard-rails) that will include letting you skip this. For now though keeping this issue open until that's completed. Good catch :)

**Comment by muellerzr** [comment]
Note that https://github.com/huggingface/accelerate/pull/1718 addresses this slightly by reducing the number of times in half

**Comment by yuxinyuan** [comment]
@muellerzr Actually, I have an idea of getting rid of calling ``get_scale()``.

Right now, the only reason we are calling ``get_scale()`` is to know if the gradients overflow or not. However, looing at the behavior of ``scaler.step()`` and ``scaler.update()``, they actually go like the following:

For ``scaler.step()``:
1. It internally invokes ``unscale_(optimizer)``
2. If no inf/NaN gradients are found, ``scaler.step(optimizer)`` runs the underlying ``optimizer.step()``. 
3. If infs/NaNs are found, ``scaler.step(optimizer)`` skips the underlying ``optimizer.step()``

For ``scaler.update()``:
4. If infs/NaNs are found (optimizer steps were skipped), ``update()`` multiplies the scale by ``backoff_factor``.
5.  If no inf/NaN gradients are found, and ``growth_interval`` unskipped iterations occurred consecutively, the scale is multiplied by ``growth_factor`` to increase it.

From 2 & 3, I think it should be possible for us to set up a ``num_step_called`` counter, and patch the optimizer's step function in a way that whenever ``step()`` is called we increment ``num_step_called``. And we only need to check the value of ``num_step_called`` to know if the gradients overflow or not

I'm not sure if I'm being clear enough. This is just my sudden inspiration, but I can make a PR on this and see if things could actually work.

**Comment by muellerzr** [comment]
If you can write a PR and some tests, feel free to pitch your idea :)

**Comment by muellerzr** [comment]
Closing with #1720
